Add PubSubSource integration test using CPS emulator

matt-kwong opened the pull request:
I added `Optional<String> emulatorEndpoint()` to `PubSubSource.builder` instead of `Optional<TransportChannelProvider>` and `Optional<CredentialsProvider>` typed params because using these types results in a `NotSerializableException` when trying to build `PubSubSource`. Note the exceptions were raised on the underlying types, not `Optional`. The implementations that raised the exceptions were `FixedTransportChannelProvider` and `NoCredentialsProvider`. I'm unsure if other implementations are serializable.

We will eventually add Docker support to enable running integration tests on CI. I'll try to work on this after learning how Apache repo CI does Dockerized testing. For now, I plan to run these tests locally.

How to run locally:

- Start the emulator by running `$ gcloud beta emulators pubsub start`
- In the `flink-connector/` directory, run `$ PUBSUB_EMULATOR_HOST=localhost:8085 mvn test -pl flink-connector-gcp-pubsub-e2e-tests -am`

clmccart commented:
in general: i dont love this pattern of changing the api to support an emulator. 

right now, it seems that the test class is responsible for checking the presence of the env variable and then setting that value on the source object. it seems more natural for the test to set the env variable and for the source object to be responsible for checking the presence of PUBSUB_EMULATOR_HOST and using it if available. it looks like this is the pattern that other pubsub clients follow: https://source.corp.google.com/piper///depot/google3/third_party/py/google/cloud/pubsub_v1/publisher/client.py;l=132?q=%20%22PUBSUB_EMULATOR_HOST%22&sq=

clmccart commented:
another general comment: do we need PubSubSourceEmulatorTest to extend PubSubEmulatorTestBase base class? it appears to me that all of the methods in PubSubEmulatorTestBase could be static helper functions and PubSubSourceEmulatorTest doesnt actually extend any implementation it just calls to those "helper" classes.  if my interpretation is correct, i think we should have something like
```
public static class PubSubEmulatorHelper {
  public static getEmulatorEndpoint () {
    . . .
  }
  . . .
}

public class PubSubSourceEmulatorTest extends TestLogger {
  . . .
  private String endpoint = PubSubEmulatorHelper.getEndpoint();
  // similar call for channel, channelProvider, etc.
  . . .
   
}
```

let me know if that makes sense

matt-kwong commented:
> in general: i dont love this pattern of changing the api to support an emulator.
> 
> right now, it seems that the test class is responsible for checking the presence of the env variable and then setting that value on the source object. it seems more natural for the test to set the env variable and for the source object to be responsible for checking the presence of PUBSUB_EMULATOR_HOST and using it if available. it looks like this is the pattern that other pubsub clients follow: https://source.corp.google.com/piper///depot/google3/third_party/py/google/cloud/pubsub_v1/publisher/client.py;l=132?q=%20%22PUBSUB_EMULATOR_HOST%22&sq=

Removed the `emulatorEndpoint` option and added the PUBSUB_EMULATOR_HOST env var check to `PubSubSource`.

matt-kwong commented on flink-connector/flink-connector-gcp-pubsub-e2e-tests/src/test/java/com/google/pubsub/flink/emulator/PubSubEmulatorTestBase.java:
This structure was copied from the existing connector: https://github.com/apache/flink-connector-gcp-pubsub/tree/main/flink-connector-gcp-pubsub-e2e-tests/src/test/java/org/apache/flink/streaming/connectors/gcp/pubsub/emulator

Do you have a preference for changing this? Maybe moving this to `flink-connector-gcp-pubsub-e2e-tests/src/main/...` or keeping this in the same folder as the e2e tests?

clmccart replied:
lets put it in the same folder as the e2e tests. that seems to be the precedent for other connectors' test clients under apache/ (https://github.com/apache/flink-connector-aws/blob/main/flink-connector-aws-e2e-tests/flink-formats-avro-glue-schema-registry-e2e-tests/src/test/java/org/apache/flink/glue/schema/registry/test/GSRKinesisPubsubClient.java)

matt-kwong replied:
Moved to the same level as the E2E test.

matt-kwong commented:
> another general comment: do we need PubSubSourceEmulatorTest to extend PubSubEmulatorTestBase base class? it appears to me that all of the methods in PubSubEmulatorTestBase could be static helper functions and PubSubSourceEmulatorTest doesnt actually extend any implementation it just calls to those "helper" classes. if my interpretation is correct, i think we should have something like

I suspect we'll eventually want tests to start the emulator instead of running it as a separate process. This is what the current connector does: https://github.com/apache/flink-connector-gcp-pubsub/blob/main/flink-connector-gcp-pubsub-e2e-tests/src/test/java/org/apache/flink/streaming/connectors/gcp/pubsub/emulator/GCloudUnitTestBase.java

I agree that tests don't need to extend from `PubSubEmulatorTestBase` given the current state, but it could be useful after making the above change. Client calls to the emulator can be moved to a helper class regardless of the above change.

Do you prefer moving everything to a helper class until the above change? Or, can this change wait until adding the above change in a future PR?

clmccart commented:
> > another general comment: do we need PubSubSourceEmulatorTest to extend PubSubEmulatorTestBase base class? it appears to me that all of the methods in PubSubEmulatorTestBase could be static helper functions and PubSubSourceEmulatorTest doesnt actually extend any implementation it just calls to those "helper" classes. if my interpretation is correct, i think we should have something like
> 
> I suspect we'll eventually want tests to start the emulator instead of running it as a separate process. This is what the current connector does: https://github.com/apache/flink-connector-gcp-pubsub/blob/main/flink-connector-gcp-pubsub-e2e-tests/src/test/java/org/apache/flink/streaming/connectors/gcp/pubsub/emulator/GCloudUnitTestBase.java
> 
> I agree that tests don't need to extend from `PubSubEmulatorTestBase` given the current state, but it could be useful after making the above change. Client calls to the emulator can be moved to a helper class regardless of the above change.
> 
> Do you prefer moving everything to a helper class until the above change? Or, can this change wait until adding the above change in a future PR?

i think i prefer moving everything to a helper class for now. if we need to extend PubSubEmulatorTestBase later, we can do that pretty easily and i suspect the functions int he helper class will be able to remain in the helper class regardless

matt-kwong commented:
> i think i prefer moving everything to a helper class for now. if we need to extend PubSubEmulatorTestBase later, we can do that pretty easily and i suspect the functions int he helper class will be able to remain in the helper class regardless

Done.
